pkgs/development/python-modules/pex/default.nix

```diff
@@ -0,0 +1,28 @@
+{lib, buildPythonPackage, fetchPypi, setuptools, wheel, requests}:
+
+buildPythonPackage rec {
+  name = "pex-${version}";
+  version = "1.4.5";
+
+  src = fetchPypi {
+    pname  = "pex";
+    sha256 = "04s9qvx87ngfs3m91qsrmk0ll16s0ldrvlxjbg4y1ic4bgsjq3hj";
+    inherit version;
+  };
+
+  propagatedBuildInputs = [ setuptools wheel requests ];
+
+  prePatch = ''
+    substituteInPlace setup.py --replace 'SETUPTOOLS_REQUIREMENT,' '"setuptools",'
```

Comment:
I'm a little worried about this line -- which was part of the original derivation. The [`setup.py` file from pex][1] includes a `pex/version.py` file, which says:
```python
# NB: If we upgrade to setuptools>=34 pex's bootstrap code in `PEXBuilder` will need an update to
# include the `packaging` package in the `.bootstrap/` code since we use
# `packaging.specifiers.SpecifierSet` - indirectly - through `pkg_resources.Requirement.specifier`.
SETUPTOOLS_REQUIREMENT = 'setuptools>=20.3,<34.0'

WHEEL_REQUIREMENT = 'wheel>=0.26.0,<0.32.0'
```

The `SETUPTOOLS_REQUIREMENT` and `WHEEL_REQUIREMENT` variables are then used in `setup.py`. When we change the requirement to just `"setuptools"`, we get whatever version is in Nix -- version 40.0.0 at the moment. This sounds dangerous in light of the above comment.

Is there a normal approach in Nix to handle situations like this? I imagine one could package setuptools version 33 and have pex use that specific version?

[1]: https://github.com/pantsbuild/pex/blob/v1.4.5/setup.py

Reply:
With python we are more bound to having only single versions in nixpkgs because of dependency propagation. However it seems that master also support setuptools newer then 34: https://github.com/pantsbuild/pex/blob/f125317bb6001ccbb56f68168bc45f947d274405/pex/version.py#L8

Reply:
Ah, cool! The new `version.py` file says:
```
# Versions 34.0.0 through 35.0.2 (last pre-36.0.0) de-vendored dependencies which
# causes problems for pex code so we exclude that range.
SETUPTOOLS_REQUIREMENT = 'setuptools>=20.3,<41,!=34.*,!=35.*'
```
which means that using version 40.0.0 should be fine. Thanks for looking this up!